plugins/inputs/azure_monitor/collector.go

```diff
@@ -0,0 +1,386 @@
+	namespace, err := getMetricsClientResponseNamespace(response)
+	if err != nil {
+		return nil, err
+	}
+
+	name, err := getMetricsClientMetricNameLocalizedValue(metric)
+	if err != nil {
+		return nil, err
+	}
+
+	replacer := strings.NewReplacer(".", "_", "/", "_", " ", "_", "(", "_", ")", "_")
+	metricName := fmt.Sprintf("azure_monitor_%s_%s",
+		replacer.Replace(strings.ToLower(*namespace)),
+		replacer.Replace(strings.ToLower(*name)))
+
+	return &metricName, nil
+}
+
+func getMetricFields(metricValues []*armmonitor.MetricValue) map[string]interface{} {
+	for index := len(metricValues) - 1; index >= 0; index-- {
+		metricFields := getMetricsClientMetricValueFields(metricValues[index])
+		if metricFields == nil {
+			continue
+		}
+
+		return metricFields
+	}
+
+	return nil
+}
+
+func getMetricTags(metric *armmonitor.Metric, response *armmonitor.MetricsClientListResponse) (map[string]string, error) {
+	tags := make(map[string]string)
+	subscriptionID, err := getMetricSubscriptionID(metric)
+	if err != nil {
+		return nil, err
+	}
+
+	tags[metricTagSubscriptionID] = *subscriptionID
+
+	resourceGroupName, err := getMetricResourceGroupName(metric)
+	if err != nil {
+		return nil, err
+	}
+
+	tags[metricTagResourceGroup] = *resourceGroupName
+
+	resourceName, err := getMetricResourceName(metric)
+	if err != nil {
+		return nil, err
+	}
+
+	tags[metricTagResourceName] = *resourceName
+
+	namespace, err := getMetricsClientResponseNamespace(response)
+	if err != nil {
+		return nil, err
+	}
+
+	tags[metricTagNamespace] = *namespace
+
+	resourceRegion, err := getMetricsClientResponseResourceRegion(response)
+	if err != nil {
+		return nil, err
+	}
+
+	tags[metricTagResourceRegion] = *resourceRegion
+
+	unit, err := getMetricsClientMetricUnit(metric)
+	if err != nil {
+		return nil, err
+	}
+
+	tags[metricTagUnit] = *unit
+
+	return tags, nil
+}
+
+func getMetricSubscriptionID(metric *armmonitor.Metric) (*string, error) {
+	metricID, err := getMetricsClientMetricID(metric)
+	if err != nil {
+		return nil, err
+	}
+
+	subscriptionID, err := getPartOfMetricID(*metricID, 0, 2, false)
+	if err != nil {
+		return nil, err
+	}
+
+	return subscriptionID, nil
+}
+
+func getMetricResourceGroupName(metric *armmonitor.Metric) (*string, error) {
+	metricID, err := getMetricsClientMetricID(metric)
+	if err != nil {
+		return nil, err
+	}
+
+	resourceGroupName, err := getPartOfMetricID(*metricID, 0, 4, false)
+	if err != nil {
+		return nil, err
+	}
+
+	return resourceGroupName, nil
+}
+
+func getMetricResourceName(metric *armmonitor.Metric) (*string, error) {
+	metricID, err := getMetricsClientMetricID(metric)
+	if err != nil {
+		return nil, err
+	}
+
+	resourceGroupName, err := getPartOfMetricID(*metricID, 1, 2, true)
+	if err != nil {
+		return nil, err
+	}
+
+	return resourceGroupName, nil
+}
+
+func getPartOfMetricID(metricID string, partIndex int, partSubPartIndex int, getPartToEnd bool) (*string, error) {
```

Comment:
Why do you work with `string` pointers in all those functions? You could simply return a string...